alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs

```diff
@@ -0,0 +1,222 @@
+    HasField "certs" (Core.TxBody era) (StrictSeq (DCert (Crypto era))),
+    HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era)))
+  ) =>
+  Core.PParams era ->
+  Tx era ->
+  UTxO era ->
+  [(AlonzoScript.Script era, [Data era], ExUnits, CostModel)]
+collectNNScriptInputs pp tx utxo =
+  let txinfo = transTx utxo tx
+   in [ (script, d : (valContext txinfo sp ++ getData tx utxo sp), eu, cost)
+        | (sp, scripthash) <- scriptsNeeded utxo tx, -- TODO, IN specification ORDER IS WRONG
+          (d, eu) <- maybeToList (indexedRdmrs tx sp),
+          script <- -- onlytwoPhaseScripts tx scripthash
+            maybeToList (Map.lookup scripthash (txscripts' (getField @"wits" tx))),
+          cost <- maybeToList (Map.lookup PlutusV1 (getField @"_costmdls" pp))
+      ]
+
+language :: Era era => AlonzoScript.Script era -> Maybe Language
+language (AlonzoScript.PlutusScript _) = Just PlutusV1
+language (AlonzoScript.TimelockScript _) = Nothing
+
+evalScripts ::
+  forall era.
+  ( Era era,
+    Alonzo.TxBody era ~ Core.TxBody era
+  ) =>
+  Tx era ->
+  [(AlonzoScript.Script era, [Data era], ExUnits, CostModel)] ->
+  Bool
+evalScripts _tx [] = True
+evalScripts tx ((AlonzoScript.TimelockScript timelock, _, _, _) : rest) =
+  evalTimelock vhks (Alonzo.vldt' (body' tx)) timelock && evalScripts tx rest
+  where
+    vhks = Set.map witKeyHash (txwitsVKey' (wits' tx))
+evalScripts tx ((AlonzoScript.PlutusScript pscript, ds, units, cost) : rest) =
+  evalPlutusScript cost units pscript (map getPlutusData ds) && evalScripts tx rest
+
+-- ===================================================================
+-- From Specification, Figure 12 "UTXOW helper functions"
+
+-- This is called checkRedeemers in the Specification
+
+-- | Check that a script has whatever associated Data that it requires.
+--     There are several ways this can happen
+--     1) The script is Not a twoPhase script, so it doesn't need any data
+--     2) The _txrdmrs Map of the TxWitness, contains Data for the script AND, either of the following
+--        a) It is not a Spending script  OR
+--        b) If it is a Spending Script, we can getData for it.
+checkScriptData ::
+  forall era.
+  ( ValidateScript era,
+    HasField "datahash" (Core.TxOut era) (StrictMaybe (DataHash (Crypto era))),
+    HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era))),
+    HasField "wdrls" (Core.TxBody era) (Wdrl (Crypto era)),
+    HasField "certs" (Core.TxBody era) (StrictSeq (DCert (Crypto era)))
+  ) =>
+  Tx era ->
+  UTxO era ->
+  (ScriptPurpose (Crypto era), ScriptHash (Crypto era)) ->
+  Bool
+checkScriptData tx utxo (sp, _h) = any ok scripts
+  where
+    scripts = txscripts' (getField @"wits" tx)
+    isSpending (Spending _) = True
+    isSpending _ = False
+    ok s =
+      isNativeScript @era s
+        || ( isJust (indexedRdmrs tx sp)
+               && (not (isSpending sp) || not (null (getData tx utxo sp)))
+           )
```

Comment:
this might be clearer written in terms of `needsData` and `hasData` predicates, and checking that if we need the data then we have it.

Reply:
done